lib/connection/file_transfer_agent/s3_util.js

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) 2015-2021 Snowflake Computing Inc. All rights reserved.
+ */
+
+const AWS = require('aws-sdk');
+const EncryptionMetadata = require('./encrypt_util').EncryptionMetadata;
+const FileHeader = require('./file_util').FileHeader;
+
+const expandTilde = require('expand-tilde');
+const fs = require('fs');
+const path = require('path');
+
+const SFC_DIGEST = 'sfc-digest';
+const AMZ_MATDESC = "x-amz-matdesc";
+const AMZ_KEY = "x-amz-key";
+const AMZ_IV = "x-amz-iv";
+const ERRORNO_WSAECONNABORTED = 10053;  // network connection was aborted
+
+const EXPIRED_TOKEN = 'ExpiredToken';
+const ADDRESSING_STYLE = 'virtual';  // explicit force to use virtual addressing style
+
+const resultStatus = require('./remote_storage_util').resultStatus;
+
+const HTTP_HEADER_VALUE_OCTET_STREAM = 'application/octet-stream';
+
+
+// S3 Location: S3 bucket name + path
+function S3Location(bucketName, s3path)
+{
+  return {
+    "bucketName": bucketName, // S3 bucket name
+    "s3path": s3path // S3 path name
+  }
+}
+
+/**
+ * Creates an S3 utility object.
+ *
+ * @returns {Object}
+ * @constructor
+ */
+function s3_util ()
+{
+  // magic number, given from  error message.
+  this.DATA_SIZE_THRESHOLD = 67108864;
+
+  this.createClient = function (stageInfo, useAccelerateEndpoint)
+  {
+    var stageCredentials = stageInfo['creds'];
+    var securityToken = stageCredentials['AWS_TOKEN'];
+
+    // if GS sends us an endpoint, it's likely for FIPS. Use it.
+    var endPoint = null;
+    if (stageInfo['endPoint'])
+    {
+      endPoint = 'https://' + stageInfo['endPoint'];
+    }
+
+    var config = {
+      region: stageInfo['region'],
+      accessKeyId: stageCredentials['AWS_KEY_ID'],
+      secretAccessKey: stageCredentials['AWS_SECRET_KEY'],
+      sessionToken: securityToken,
+      endpoint: endPoint,
+      useAccelerateEndpoint: useAccelerateEndpoint
+    };
+
+    AWS.config.update(config);
+
+    var s3 = new AWS.S3({
+      apiVersion: '2006-03-01',
+      signatureVersion: 'v4'
+    });
+
+    return s3;
+  }
+
+  this.extractBucketNameAndPath = function (stageLocation)
+  {
+    // expand '~' and '~user' expressions
+    stageLocation = expandTilde(stageLocation);
+
+    var bucketName = stageLocation;
+    var s3path;
+
+    // split stage location as bucket name and path
+    if (stageLocation.includes('/'))
+    {
+      bucketName = stageLocation.substring(0, stageLocation.indexOf('/'));
+
+      s3path = stageLocation.substring(stageLocation.indexOf('/') + 1, stageLocation.length);
+      if (s3path && !s3path.endsWith('/'))
+      {
+        s3path += '/';
+      }
+    }
+    return S3Location(bucketName, s3path);
+  }
+
+  this.getFileHeader = async function (meta, filename)
+  {
+    var stageInfo = meta['stageInfo'];
+    var client = this.createClient(stageInfo);
+    var s3location = this.extractBucketNameAndPath(stageInfo['location']);
+
+    var params = {
+      Bucket: s3location.bucketName,
+      Key: filename
+    };
```

Comment:
In the PythonConnector we don't use `filename` directly, we do this:

`s3path = s3location.s3path + filename.lstrip("/")`

And set that as the key. This could be why you're getting permission errors